Restore bootstrap pre-release support

marcparadise opened the pull request:
## Description

Bootstrap prerelease support was originally removed after
 discussion in community slack, but it turns out to have a little more 
utility than we thought at the time. When enabled, it will cause the 
package to be downloaded from the `current` channel - the latest by default,
but otherwise the specified `--bootstrap-version`.

`Knife#config_source` has been updated to not use a separate hash to store config sources,
and now knows about `:cli_default` value source, which maps to defaulted CLI options that are not in config or used on the CLI by the operator.  This was necessary to prevent `knife bootstrap` from thinking that `--channel` had been set by the operator when it only had a default value.

The CLI flag is now `--channel CHANNEL`; `--prerelease` has been deprecated. If use, will set `channel` to "current". Default is `stable`.

This PR otherwise restores the original behavior (minus the bits that were specific to gem prerelease)  and adds some tests.



Signed-off-by: Marc A. Paradise <marc.paradise@gmail.com>

## Related Issue
<!--- If you are suggesting a new feature or change, please create an issue first -->
<!--- Please link to the issue, discourse, or stackoverflow here: -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

lamont-granquist commented:
i think the way it is trying to do this is very, very old though isn't it?

the current channel also won't have any prerelease gem versions in it, so the autodetection of the prerelease gem version is kind of dead code.

and then don't we really want to be passing in the channel option?  i don't see where you're passing in the current channel...

btm commented on lib/chef/knife/core/bootstrap_context.rb:
```diff
@@ -190,13 +190,24 @@ def knife_config
+            # If bootstrapping a pre-release version add -p to the installer string
+            if chef_version_string.split(".").length > 3
+              installer_version_string << "-p"
+            end
```
We no longer make "prerelease" gems (x.y.z.pre.0, x.y.z.beta.0) and just use channels now, so this conditional can be dropped.

tas50 commented on lib/chef/knife/bootstrap.rb:
```diff
@@ -146,6 +146,10 @@ class Bootstrap < Knife
         description: "The version of Chef to install",
         proc: lambda { |v| Chef::Config[:knife][:bootstrap_version] = v }
 
+      option :prerelease,
```
any reason not to just expose a channel flag. That's a term people would be familiar with and we'd avoid any sort of magic by just passing in the channel name. That would also make it future proof if we added more channels in the future.

btm replied:
Agree with @tas50, let's add a "channel" option. Then instead of adding `-p` to the installer_version string (which is just passed to install.sh),  we would pass `-c CHANNEL`.

The prerelease option could warn it is deprecated and set the channel to "current" which is functionally our prerelease builds now.

marcparadise replied:
Cool, I'd had the same thought and figured few changes are better at this point.

btm replied:
it's a small change that's overdue. I think it will also let us install from unstable which will let us get the license acceptance builds that we'll start making soon that we're going to intentionally keep from hitting current for a couple days.

btm commented on test.rb:
```diff
@@ -0,0 +1,26 @@
+def rl
```
I don't think you meant to add this file?

marcparadise replied:
Nope, thanks

marcparadise commented:
I have updated this branch with the changes above, however because I needed to touch deprecations I rebased it on top of #8429 which also includes deprecation work. way.

btm commented on lib/chef/knife/bootstrap.rb:
```diff
@@ -146,6 +147,12 @@ class Bootstrap < Knife
+        long: "--channel CHANNEL",
+        description: "Install from the given channel. Valid values are 'current' and 'stable'. Default is 'stable'",
+        default: "stable",
+        in: %w{stable current}
```
Can we allow `unstable` here too? There are testing scenarios where we would want the ease of grabbing a build from that channel and it is accessible.

btm commented on lib/chef/knife/core/bootstrap_context.rb:
```diff
@@ -187,16 +187,32 @@ def knife_config
+                      # We will take the latest current by default,
+                      # if no specific version is given.
+                      if use_current_channel
+                        # Take the latest stable version from the current major release.
```
> # Take the latest stable version from the latest major release.
or highest?

btm commented on lib/chef/knife/core/bootstrap_context.rb:
```diff
@@ -187,16 +187,32 @@ def knife_config
+          # NOTE: Changes here should also be reflected in Knife::Core::BootstrapContext#latest_current_chef_version_string
+          installer_version_string = []
+          use_current_channel = (@config[:channel] == "current")
+          installer_version_string << "-p" if use_current_channel
```
`-p` is actually back-compat in `install.sh`, we should pass `-c CHANNEL` install the installer_version_string here.

It is easiest to see here: https://omnitruck.chef.io/install.sh

But it comes from here: https://github.com/chef/mixlib-install/blob/master/lib/mixlib/install/generator/bourne/scripts/script_cli_parameters.sh

btm commented on lib/chef/knife/core/windows_bootstrap_context.rb:
```diff
@@ -158,14 +158,35 @@ def start_chef
         def latest_current_windows_chef_version_query
+          # NOTE: Changes here should also be reflected in Knife::Core::BootstrapContext#latest_current_chef_version_string
+          use_current_channel = (@config[:channel] == "current")
+          installer_version_string = use_current_channel ? "&prerelease=true" : ""
```
fixing this to pass a channel instead of just prerelease means changing the URL down in `msi_url` to use omnitruck instead of whatever that chef.io redirect is.

https://docs.chef.io/api_omnitruck.html

I don't know how much time we want to spend on that though, because what we really need to do is replace all this with https://github.com/chef/mixlib-install

tas50 commented on lib/chef/knife/core/bootstrap_context.rb:
```diff
@@ -187,16 +187,17 @@ def knife_config
+          return knife_config[:bootstrap_version] if knife_config[:bootstrap_version]
+
+          if @config[:channel] == "stable"
+            Chef::VERSION.split(".").first
```
Yet another reason we need to use mixlib-install

btm commented:
failure is opensuseleap42, merging.
